### PR TITLE
Refactor render_kernel for NV [run_process_replay]

### PR DESCRIPTION
**Pull request by ignaciosica**
It's failing process_replay because it is no longer rendering the vectorization structs if there are not used. Before, if there was type half or bfloat present, the vectorization structs would render automatically.

**kernel failing process_replay**

```c
#define INFINITY (__int_as_float(0x7f800000))
#define NAN (__int_as_float(0x7fffffff))
#include <cuda_fp16.h>
extern "C" __global__ void __launch_bounds__(1) r_4096_32000_32000n1(half* data0, const int* data1, const half* data2) {
  int gidx0 = blockIdx.x; /* 4096 */
  int val0 = data1[0];
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 32000; ridx0++) {
    half val1 = data2[gidx0+(ridx0*4096)];
    acc0 = (acc0+(float)(((half)((!(ridx0!=val0)))*val1)));
  }
  data0[gidx0] = (half)(acc0);
}
```

``` diff
 #define INFINITY (__int_as_float(0x7f800000))
 #define NAN (__int_as_float(0x7fffffff))
 #include <cuda_fp16.h>
-struct __align__(8) half4 { half x, y, z, w; }; __device__ half4 make_half4(half x, half y, half z, half w) { half4 r={x, y, z, w}; return r; }
-struct __align__(16) half8 { half x, y, z, w, a, b, c, d; }; __device__ half8 make_half8(half x, half y, half z, half w, half a, half b, half c, half d) { half8 r={x, y, z, w, a, b, c, d}; return r; }
 extern "C" __global__ void __launch_bounds__(1) r_4096_32000_32000(half* data0, const int* data1, const half* data2) {
   int gidx0 = blockIdx.x; /* 4096 */
   int val0 = data1[0];
Process Process-16:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/runner/work/tinygrad/tinygrad/process_replay.py", line 69, in diff_kernel
    if ASSERT_DIFF: raise e
```

### CODE REVIEW

File: tinygrad/renderer/cstyle.py [hunk 1]
```diff
@@ -268,9 +268,10 @@ def render_kernel(self, function_name, kernel, bufs, uops, prefix=None):
                     UnaryOps.EXP2: lambda x,dtype: f"hexp2({x})" if dtype in (dtypes.half, dtypes.bfloat16) else f"exp2({x})",}
 
 _nms = "xyzwabcdefghijkl"
-def _make_cuda_dtype(base_type, cnt, align):
-  vec, elems, header = f"{base_type}{cnt}", ', '.join(_nms[:cnt]), ', '.join([f"{base_type} {x}" for x in _nms[:cnt]])
-  return f"struct __align__({align}) {vec} {{ {base_type} {elems}; }}; __device__ {vec} make_{vec}({header}) {{ {vec} r={{{elems}}}; return r; }}"
+def _make_cuda_dtype(self, dtype):
+  vec, scalar = self.render_dtype(dtype), self.render_dtype(dtype.scalar()),
+  elems, header = ', '.join(_nms[:dtype.count]), ', '.join([f"{scalar} {x}" for x in _nms[:dtype.count]])
+  return f"struct __align__({dtype.itemsize}) {vec} {{ {scalar} {elems}; }}; __device__ {vec} make_{vec}({header}) {{ {vec} r={{{elems}}}; return r; }}" # noqa:E501
```
Comment (Qazalin) [inline tinygrad/renderer/cstyle.py:274]: no new noqa

File: tinygrad/renderer/cstyle.py [hunk 2]
```diff
@@ -268,9 +268,10 @@ def render_kernel(self, function_name, kernel, bufs, uops, prefix=None):
                     UnaryOps.EXP2: lambda x,dtype: f"hexp2({x})" if dtype in (dtypes.half, dtypes.bfloat16) else f"exp2({x})",}
 
 _nms = "xyzwabcdefghijkl"
-def _make_cuda_dtype(base_type, cnt, align):
-  vec, elems, header = f"{base_type}{cnt}", ', '.join(_nms[:cnt]), ', '.join([f"{base_type} {x}" for x in _nms[:cnt]])
-  return f"struct __align__({align}) {vec} {{ {base_type} {elems}; }}; __device__ {vec} make_{vec}({header}) {{ {vec} r={{{elems}}}; return r; }}"
+def _make_cuda_dtype(self, dtype):
```
Comment (Qazalin) [inline tinygrad/renderer/cstyle.py:271]: can you type-check this, also why self? This isn't a class method.

File: tinygrad/renderer/cstyle.py [hunk 3]
```diff
@@ -293,19 +294,15 @@ def __init__(self, arch:str): self.tensor_cores = CUDARenderer.tensor_cores if i
...
 
-    if any(uop.dtype == dtypes.bfloat16 for uop in uops):
-      prefix += ["#include <cuda_bf16.h>"] + [_make_cuda_dtype("nv_bfloat16", x, x*2) for x in [4, 8]]
+    for dtype in dedup(uop.dtype for uop in uops if uop.dtype is not None and uop.dtype.scalar() in (dtypes.half, dtypes.bfloat16)):
+      prefix += [f"#include <cuda_{'fp' if dtype == dtypes.half else 'bf'}16.h>"] if dtype.count == 1 else [_make_cuda_dtype(self, dtype)]
 
     # TODO: this has to be way better to generate for arbitrary M,N,K: use arg[1] for MNK, use arg[4] for vec sizes, encode register packing
     for arg in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
-      fn, ti, to, ci, co = arg[0], dt_map[arg[2]][0], dt_map[arg[3]][0], dt_map[arg[2]][1], dt_map[arg[3]][1]
+      dt_map = { dtypes.float: "f32", dtypes.half: "f16", dtypes.bfloat16: "bf16" }
```
Comment (Qazalin) [inline tinygrad/renderer/cstyle.py:304]: why is dt_map inside the loop?

File: tinygrad/renderer/cstyle.py [hunk 4]
```diff
@@ -293,19 +294,15 @@ def __init__(self, arch:str): self.tensor_cores = CUDARenderer.tensor_cores if i
...
-    dt_map = { dtypes.float: ("float","f32"), dtypes.half: ("half","f16"), dtypes.bfloat16: ("nv_bfloat16","bf16"), }
-
     prefix = ["#define INFINITY (__int_as_float(0x7f800000))","#define NAN (__int_as_float(0x7fffffff))"]
-    if any(uop.dtype == dtypes.half for uop in uops):
-      prefix += ["#include <cuda_fp16.h>"] + [_make_cuda_dtype("half", x, x*2) for x in [4, 8]]
 
-    if any(uop.dtype == dtypes.bfloat16 for uop in uops):
-      prefix += ["#include <cuda_bf16.h>"] + [_make_cuda_dtype("nv_bfloat16", x, x*2) for x in [4, 8]]
+    for dtype in dedup(uop.dtype for uop in uops if uop.dtype is not None and uop.dtype.scalar() in (dtypes.half, dtypes.bfloat16)):
+      prefix += [f"#include <cuda_{'fp' if dtype == dtypes.half else 'bf'}16.h>"] if dtype.count == 1 else [_make_cuda_dtype(self, dtype)]
```
Comment (Qazalin) [inline tinygrad/renderer/cstyle.py:300]: let's separate the dtype.count conditional to another diff without [run_process_replay].

Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: ok

File: tinygrad/renderer/cstyle.py (hunk 3 above)
Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: should be above, sorry

File: tinygrad/renderer/cstyle.py (hunk 2 above)
Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: you are right, fixed

File: tinygrad/renderer/cstyle.py (hunk 1 above)
Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: fixed

**Comment by Qazalin** [comment]
cool, I'm supportive of adding `render_vector_prefix` to the CUDA and HIP renderers like this:
image

**Comment by ignaciosica** [comment]
> cool, I'm supportive of adding `render_vector_prefix` to the CUDA and HIP renderers

Sure, refactoring it later today. I don't feel completely comfortable with AMD though because I don't have access to an AMD machine. Can I completely trust CI with process_replay for this refactoring?

**Comment by Qazalin** [comment]
yeah CI works for this, I'll push to real benchmarks too once you have it ready.

**Comment by ignaciosica** [comment]
Ok, thanks

**Pull request merged**